Zend/zend_operators.c

```diff
@@ -70,14 +70,12 @@ static _locale_t current_locale = NULL;
 
 #define BLOCKCONV_LOAD(input) \
 	__m128i blconv_operand = _mm_loadu_si128((__m128i*)(input)); \
-	__m128i blconv_gt = _mm_cmpgt_epi8(blconv_operand, blconv_start_minus_1); \
-	__m128i blconv_lt = _mm_cmplt_epi8(blconv_operand, blconv_end_plus_1); \
-	__m128i blconv_mingle = _mm_and_si128(blconv_gt, blconv_lt);
+	__m128i blconv_mask = _mm_cmplt_epi8(_mm_add_epi8(blconv_operand, blconv_offset), blconv_threshold);
```

Comment:
Just to check (I'm not super familiar with SIMD), `_mm_add_epi8` has well-defined wrap-around behavior?

Reply:
I'd also find `_mm_sub_epi8` and then just using `start` for `blconv_offset` more obvious. Or is `_mm_add_epi8` faster?

Reply:
@iluuu1994 I would have to check the manual, but in this case, there will never be any wrap-around. That is because the 'offset' is only added to bytes which are lowercase/uppercase ASCII characters, and the offset is just big enough to change lowercase characters to uppercase or vice versa.

Using `start` for the offset wouldn't work. The offset must be the difference between a lowercase ASCII character and the corresponding uppercase character, like 'A' and 'a'.

Reply:
If we use `_mm_sub_epi8`, the code above would become `offset = _mm_set1_epi8((char)(start + 128));`. Do you feel that would read better?

Reply:
> That is because the 'offset' is only added to bytes which are lowercase/uppercase ASCII characters, and the offset is just big enough to change lowercase characters to uppercase or vice versa.

Hmm, but we're doing `_mm_add_epi8` on the entire `blconv_operand` which might contain any characters from `CHAR_MIN` to `CHAR_MAX`, right? Aren't we even depending on the fact that there's wrap-around to push them over the upper bounds so we can used `cmplt`? Sorry if I'm missing the point.

> If we use _mm_sub_epi8, the code above would become offset = _mm_set1_epi8((char)(start + 128));. Do you feel that would read better?

Oh, right, so that's not really much better than, let's leave it as is then.

Reply:
@iluuu1994 Sigh... another incorrect comment. When I said there will be no wrap-around, I meant when adding `blconv_delta` there will be no wrap-around. When adding `blconv_offset`, there will definitely be wrap-around. Sorry for the confusion.

Reply:
> > That is because the 'offset' is only added to bytes which are lowercase/uppercase ASCII characters, and the offset is just big enough to change lowercase characters to uppercase or vice versa.
> 
> Hmm, but we're doing `_mm_add_epi8` on the entire `blconv_operand` which might contain any characters from `CHAR_MIN` to `CHAR_MAX`, right? Aren't we even depending on the fact that there's wrap-around to push them over the upper bounds so we can used `cmplt`? Sorry if I'm missing the point.

You are totally right, I was confused.

Just checking the manual.

Reply:
OK, here we are, *Intel 64 and IA-32 Architectures Software Developer's Manual*, Sept 2014 edition, page 1134.

> PADDB/PADDW/PADDD—Add Packed Integers

> Adds the packed byte, word, doubleword, or quadword integers in the first source operand to the second source operand and stores the result in the destination operand. When a result is too large to be represented in the 8/16/32 integer (overflow), the result is wrapped around and the low bits are written to the destination element (that is, the carry is ignored).

Does that sound good?

Reply:
Perfect, thank you @alexdowad!